// @oliverbill/fastroute-domain/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ALLOWED_WAYPOINT_CURRENT_STATUSES: () => ALLOWED_WAYPOINT_CURRENT_STATUSES,
  ALLOWED_WAYPOINT_TARGET_STATUSES: () => ALLOWED_WAYPOINT_TARGET_STATUSES,
  ROUTE_FINISH_REQUIRED_STATUS: () => ROUTE_FINISH_REQUIRED_STATUS,
  canFinishRoute: () => canFinishRoute,
  dbscanDefaults: () => dbscanDefaults,
  err: () => err,
  generateWaypointsForCluster: () => generateWaypointsForCluster,
  groupByClusterId: () => groupByClusterId,
  isAllowedWaypointCurrentStatus: () => isAllowedWaypointCurrentStatus,
  isAllowedWaypointTargetStatus: () => isAllowedWaypointTargetStatus,
  normalizeJwtProviderResult: () => normalizeJwtProviderResult,
  ok: () => ok,
  validateDbscanPoints: () => validateDbscanPoints,
  validateFinishWaypoint: () => validateFinishWaypoint,
  validateJwtInput: () => validateJwtInput,
  validateWaypointPhotoMeta: () => validateWaypointPhotoMeta
});
module.exports = __toCommonJS(index_exports);

// src/errors.ts
function err(error, code = "INVALID_INPUT", details) {
  if (details === void 0) {
    return { ok: false, error, code };
  }
  return { ok: false, error, code, details };
}
function ok(value, warnings) {
  if (warnings === void 0) {
    return { ok: true, value };
  }
  return { ok: true, value, warnings };
}

// src/auth.ts
var import_zod = require("zod");
function validateJwtInput(input) {
  const schema = import_zod.z.object({
    authorization: import_zod.z.string().min(1)
  });
  const parsed = schema.safeParse({ authorization: input.authorization ?? "" });
  if (!parsed.success) {
    return err("Authorization token ausente.", "UNAUTHORIZED");
  }
  const token = parsed.data.authorization.trim();
  return ok({ token });
}
function normalizeJwtProviderResult(res) {
  const id = (res.id ?? "")?.toString().trim();
  if (!id) {
    return err(res.error?.toString() || "JWT inv\xE1lido ou expirado.", "UNAUTHORIZED");
  }
  return ok({ userId: id });
}

// src/cluster.ts
var import_zod2 = require("zod");
function dbscanDefaults(partial) {
  const eps = typeof partial?.eps === "number" && Number.isFinite(partial.eps) ? partial.eps : 1e-3;
  const minPts = typeof partial?.minPts === "number" && Number.isFinite(partial.minPts) && partial.minPts > 0 ? Math.floor(partial.minPts) : 2;
  return { eps, minPts };
}
function validateDbscanPoints(points) {
  const schema = import_zod2.z.array(
    import_zod2.z.object({
      address_id: import_zod2.z.number().int().positive(),
      lat: import_zod2.z.number().finite(),
      longitude: import_zod2.z.number().finite()
    })
  ).min(2);
  const parsed = schema.safeParse(points);
  if (!parsed.success) {
    return err(
      "Coordenadas insuficientes ou inv\xE1lidas para gerar cluster (m\xEDnimo 2 pontos v\xE1lidos).",
      "INVALID_INPUT",
      { issues: parsed.error.issues }
    );
  }
  return ok(parsed.data);
}
function groupByClusterId(items) {
  const grouped = {};
  for (const it of items) {
    const cid = it.cluster_id;
    grouped[cid] ||= [];
    grouped[cid].push(it);
  }
  return grouped;
}

// src/routes.ts
var ROUTE_FINISH_REQUIRED_STATUS = "EM_ANDAMENTO";
function canFinishRoute(input) {
  const route = input.route;
  if (!route) {
    return err("Rota n\xE3o encontrada.", "NOT_FOUND");
  }
  if (route.status !== ROUTE_FINISH_REQUIRED_STATUS) {
    return err("Nao \xE9 poss\xEDvel finalizar a rota! Ela n\xE3o est\xE1 em andamento!", "INVALID_STATE", {
      expected: ROUTE_FINISH_REQUIRED_STATUS,
      actual: route.status
    });
  }
  const hasPending = input.waypoints.some((w) => w.status === "PENDENTE");
  if (hasPending) {
    return err("N\xE3o \xE9 Poss\xEDvel finalizar a rota. Existem Entregas Pendentes !", "INVALID_STATE");
  }
  return ok(true);
}

// src/waypoints.ts
var import_zod3 = require("zod");
var ALLOWED_WAYPOINT_TARGET_STATUSES = [
  "ENTREGUE",
  "FALHA MORADOR AUSENTE",
  "FALHA TEMPO ADVERSO"
];
var ALLOWED_WAYPOINT_CURRENT_STATUSES = ["PENDENTE", "REORDENADO"];
function isAllowedWaypointTargetStatus(status) {
  return ALLOWED_WAYPOINT_TARGET_STATUSES.includes(status);
}
function isAllowedWaypointCurrentStatus(status) {
  return ALLOWED_WAYPOINT_CURRENT_STATUSES.includes(status);
}
function generateWaypointsForCluster(input) {
  const schema = import_zod3.z.object({
    route_id: import_zod3.z.number().int().positive(),
    cluster_id: import_zod3.z.number().int(),
    addresses: import_zod3.z.array(
      import_zod3.z.object({
        address_id: import_zod3.z.number().int().positive()
      })
    ).min(1)
  });
  const parsed = schema.safeParse(input);
  if (!parsed.success) {
    return err("Dados inv\xE1lidos para gerar waypoints.", "INVALID_INPUT", { issues: parsed.error.issues });
  }
  const { route_id, cluster_id, addresses } = parsed.data;
  const out = addresses.map((a, idx) => ({
    route_id,
    cluster_id,
    address_id: a.address_id,
    seq_order: idx + 1,
    status: "PENDENTE"
  }));
  return ok(out);
}
function validateWaypointPhotoMeta(photo) {
  const schema = import_zod3.z.object({
    waypoint_id: import_zod3.z.union([import_zod3.z.number().int().positive(), import_zod3.z.string().min(1)]),
    filename: import_zod3.z.string().min(1),
    user_id: import_zod3.z.union([import_zod3.z.number().int().positive(), import_zod3.z.string().min(1)]),
    object_path: import_zod3.z.string().min(1),
    file_size_bytes: import_zod3.z.union([import_zod3.z.number().positive(), import_zod3.z.string().min(1)]),
    photo_url: import_zod3.z.string().min(1)
  });
  const parsed = schema.safeParse(photo);
  if (!parsed.success) {
    return err("Campos obrigat\xF3rios n\xE3o preenchidos (foto).", "MISSING_FIELDS", { issues: parsed.error.issues });
  }
  return ok(parsed.data);
}

// src/validation.ts
function validateFinishWaypoint(input) {
  const current = input.currentWaypoint;
  if (!current) {
    return err("waypoint nao encontrado!", "NOT_FOUND");
  }
  if (!ALLOWED_WAYPOINT_TARGET_STATUSES.includes(input.targetStatus)) {
    return err("Status deve ser ENTREGUE ou FALHA!", "INVALID_TRANSITION", {
      allowedTargets: ALLOWED_WAYPOINT_TARGET_STATUSES,
      got: input.targetStatus
    });
  }
  if (!ALLOWED_WAYPOINT_CURRENT_STATUSES.includes(current.status)) {
    return err("Status atual do waypoint deve ser PENDENTE!", "INVALID_STATE", {
      allowedCurrents: ALLOWED_WAYPOINT_CURRENT_STATUSES,
      got: current.status
    });
  }
  const photoOk = validateWaypointPhotoMeta(input.photo);
  if (!photoOk.ok) {
    return photoOk;
  }
  return ok({
    waypoint_id: current.id,
    new_status: input.targetStatus,
    obs_falha: input.obs_falha ?? null
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ALLOWED_WAYPOINT_CURRENT_STATUSES,
  ALLOWED_WAYPOINT_TARGET_STATUSES,
  ROUTE_FINISH_REQUIRED_STATUS,
  canFinishRoute,
  dbscanDefaults,
  err,
  generateWaypointsForCluster,
  groupByClusterId,
  isAllowedWaypointCurrentStatus,
  isAllowedWaypointTargetStatus,
  normalizeJwtProviderResult,
  ok,
  validateDbscanPoints,
  validateFinishWaypoint,
  validateJwtInput,
  validateWaypointPhotoMeta
});
